Engine/Helper.cs

```diff
@@ -3555,4 +3555,239 @@ public object VisitUsingExpression(UsingExpressionAst usingExpressionAst)
             return null;
         }
     }
+
+    /// Class to represent a directed graph
+    public class Digraph<T>
+    {
+        private List<List<int>> graph;
+        private Dictionary<T, int> vertexIndexMap;
+
+        /// <summary>
+        /// Public constructor
+        /// </summary>
+        public Digraph()
+        {
+            graph = new List<List<int>>();
+            vertexIndexMap = new Dictionary<T, int>();
+        }
+
+        /// <summary>
+        /// Construct digraph with an EqualityComparer used for comparing type T
+        /// </summary>
+        /// <param name="equalityComparer"></param>
+        public Digraph(IEqualityComparer<T> equalityComparer) : this()
+        {
+            if (equalityComparer == null)
+            {
+                throw new ArgumentNullException("equalityComparer");
+            }
+
+            vertexIndexMap = new Dictionary<T, int>(equalityComparer);
+        }
+
+        /// <summary>
+        /// Return the number of vertices in the graph
+        /// </summary>
+        public int NumVertices
+        {
+            get { return graph.Count; }
+        }
+
+        /// <summary>
+        /// Return an enumerator over the vertices in the graph
+        /// </summary>
+        public IEnumerable<T> GetVertices()
+        {
+            return vertexIndexMap.Keys;
+        }
+
+        /// <summary>
+        /// Check if the given vertex is part of the graph.
+        ///
+        /// If the vertex is null, it will throw an ArgumentNullException.
+        /// If the vertex is non-null but not present in the graph, it will throw an ArgumentOutOfRangeException
```

Comment:
Line 3608 is not applicable for this method (`ContainsVertex`). I will remove it. The method `ContainsVertex` expects a non-null object for comparison and returns a `bool` type.